giftcertificate/dao/src/main/java/com/epam/esm/dao/impl/GiftCertificateJDBCTemplate.java

```diff
@@ -0,0 +1,227 @@
+     * Read GiftCertificate from DB by id
+     *
+     * @param id long type parameter
+     * @return Optional<GiftCertificate>
+     * @throws IdNotExistDAOException if records with such id not exist in DB
+     */
+    @Override
+    public GiftCertificate read(long id) throws IdNotExistDAOException {
+        Optional<GiftCertificate> giftCertificate = jdbcTemplate.query(SELECT_CERTIFICATE_ID, new Object[]{id}, giftCertificateMapper).stream().findFirst();
+        if (!giftCertificate.isPresent()) {
+            logger.error("GiftCertificate with id=" + id + " is not exist in DB");
+            throw new IdNotExistDAOException("There is no gift certificate with ID = " + id + " in Database");
+        }
+        List<Tag> tags = jdbcTemplate.query(SELECT_TAGS_BY_CERTIFICATE_ID, new Object[]{id}, tagMapper);
+        logger.info("GiftCertificate is read from DB");
+        for (Tag tag : tags) {
+            giftCertificate.get().getTags().add(tag);
+        }
+        return giftCertificate.get();
+    }
+
+    /**
+     * Update GiftCertificate
+     *
+     * @param giftCertificate we wont update
+     * @return updated GiftCertificate
+     */
+    @Override
+    public GiftCertificate update(GiftCertificate giftCertificate) throws UpdateDAOException {
+        int i = jdbcTemplate.update(UPDATE_GIFT_CERTIFICATE,
+                giftCertificate.getName(),
+                giftCertificate.getDescription(),
+                giftCertificate.getPrice(),
+                giftCertificate.getDuration(),
+                LocalDateTime.now(ZoneId.systemDefault()),
+                giftCertificate.getId());
+        if (i == 0) {
+            logger.info("GiftCertificate is not updated in DB");
+            throw new UpdateDAOException("giftCertificate has not been updated");
+        }
+        logger.info("GiftCertificate has been updated in DB");
+        return giftCertificate;
+    }
+
+    /**
+     * Delete GiftCertificate from DB by id
+     *
+     * @param id GiftCertificate with this id will be deleted from DB
+     */
+    @Override
+    public void delete(long id) throws IdNotExistDAOException {
+        int i = jdbcTemplate.update(DELETE_GIFT_CERTIFICATE, id);
+        if (i == 0) {
+            logger.info("GiftCertificate isn't deleted from DB");
+            throw new IdNotExistDAOException("GiftCertificate with id = " + id + "is not exist in DB");
+        } else {
+            logger.info("GiftCertificate is deleted from DB");
+        }
+    }
+
+    /**
+     * Find all giftCertificates with condition determined by parameters
+     *
+     * @param sortType  name of field of table in DB
+     * @param orderType ASC or DESC
+     * @return list og GiftCertificates
+     * @throws RequestParamDAOException if parameters don't right
+     */
+    @Override
+    public List<GiftCertificate> findAll(String sortType, String orderType) throws RequestParamDAOException {
+        List<GiftCertificate> giftCertificateList;
+        try {
```

Comment:
logic to manager, your dao should get calculated sortType and orderType from manager